src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/PackageRestore/PackageRestoreDataSource.cs

```diff
@@ -115,34 +145,86 @@ internal async Task<IEnumerable<IProjectVersionedValue<RestoreData>>> RestoreAsy
             };
         }
 
-        private async Task<bool> RestoreCoreAsync(ProjectRestoreInfo restoreInfo)
+        private async Task<bool> RestoreCoreAsync(PackageRestoreUnconfiguredInput value)
         {
-            // Restore service always does work regardless of whether the value we pass 
-            // them to actually contains changes, only nominate if there are any.
-            byte[] hash = RestoreHasher.CalculateHash(restoreInfo);
+            ProjectRestoreInfo? restoreInfo = value.RestoreInfo;
+            bool success = false;
 
-            if (_latestHash != null && Enumerable.SequenceEqual(hash, _latestHash))
-                return true;
+            try
+            {
+                // Prevent extra restore under some conditions.
+                if (IsProjectConfigurationVersionOutOfDate(value.ConfiguredInputs))
+                {
+                    return false;
```

Comment:
This false is [going to cause](http://index/?leftProject=Microsoft.VisualStudio.ProjectSystem.Managed.VS&leftSymbol=k7cwyziacx79&file=ProjectSystem%5CVS%5CPackageRestore%5CPackageRestoreProgressTracker.PackageRestoreProgressTrackerInstance.cs&line=101) restore to be considered up-to-date and operation progress will be considered finished. I don't think this is what we want, I think this should probably be true. @lifengl

Reply:
So, a failed restore will make the operation progress to finish?  I didn't check the logic in down chain dataflow, and it is good that you reviewed it.

no, i don't feel return 'true' is correct either.  I think we need move this check logic into RestoreAsync before calling RestoreCoreAsync, and when this condition matches, RestoreAsync return an empty array, which will skip the data, and not push it to the next dataflow block.

Reply:
Change done